gobblin-utility/src/main/java/gobblin/util/ExecutorsUtils.java

```diff
@@ -146,4 +155,73 @@ public static void shutdownExecutorService(ExecutorService executorService, Opti
     shutdownExecutorService(executorService, logger, EXECUTOR_SERVICE_SHUTDOWN_TIMEOUT,
         EXECUTOR_SERVICE_SHUTDOWN_TIMEOUT_TIMEUNIT);
   }
+
+  /**
+   * A utility method to parallelize loops. Applies the {@link Function} to every element in the {@link List} in
+   * parallel by spawning threads. A list containing the result obtained by applying the function is returned. The
+   * method is a blocking call and will wait for all the elements in the list to be processed or timeoutInSecs which
+   * ever is earlier.
+   * <p>
+   * <b>NOTE: The method is an all or none implementation. Meaning, if any of the thread fails, the method will throw an
+   * {@link ExecutionException} even if other threads completed successfully</b>
+   * </p>
+   *
+   * <ul>
+   * <li>Uses a Fixed thread pool of size threadCount.
+   * <li>Uses {@link #shutdownExecutorService(ExecutorService, Optional, long, TimeUnit)} to shutdown the executor
+   * service
+   * <li>All threads are daemon threads
+   * </ul>
+   *
+   * @param list input list on which the function is applied in parallel
+   * @param function to be applied on every element of the list
+   * @param threadCount to be used to process the list
+   * @param timeoutInSecs to wait for all the threads to complete
+   * @param logger an {@link Optional} wrapping a {@link Logger} to be used during shutdown
+   *
+   * @return a list containing the result obtained by applying the function on each element of the input list in the
+   *         same order
+   *
+   * @throws IllegalArgumentException if input list or function is null
+   * @throws ExecutionException <ul>
+   *           <li>if any computation threw an exception
+   *           <li>if any computation was cancelled
+   *           <li>if any thread was interrupted while waiting
+   *           <ul>
+   */
+  public static <T, F> List<T> parallelize(final List<F> list, final Function<F, T> function, int threadCount,
+      int timeoutInSecs, Optional<Logger> logger) throws ExecutionException {
+
+    Preconditions.checkArgument(list != null, "Input list can not be null");
+    Preconditions.checkArgument(function != null, "Function can not be null");
+
+    final List<T> results = Lists.newArrayList();
```

Comment:
nit: This is a small optimization. I'd probably use "new ArrayList<T>(list.size())". Thus we can avoid the dynamic growing of the list as we already now the final size.